### PR TITLE
fix(format/html): preserve whitespace between a HtmlSingleTextExpression/HtmlElement and HtmlContent

### LINKED ISSUE #8584: 📝 HTML formatter removes space from rendered content after variable in Svelte
### Environment information

```bash
CLI:
  Version:                      2.3.10
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           linux

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  BIOME_THREADS:                unset
  NO_COLOR:                     unset
  TERM:                         xterm-256color
  JS_RUNTIME_VERSION:           v25.2.1
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         pnpm/10.25.0

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         biome.json
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  true
  HTML full support enabled:    true

Formatter:
  Format with errors:           unset
  Indent style:                 Tab
  Indent width:                 unset
  Line ending:                  unset
  Line width:                   unset
  Attribute position:           unset
  Bracket spacing:              unset
  Includes:                     unset

JavaScript Formatter:
  Enabled:                      unset
  JSX quote style:              unset
  Quote properties:             unset
  Trailing commas:              unset
  Semicolons:                   unset
  Arrow parentheses:            unset
  Bracket spacing:              unset
  Bracket same line:            unset
  Quote style:                  Double Quotes
  Indent style:                 unset
  Indent width:                 unset
  Line ending:                  unset
  Line width:                   unset
  Attribute position:           unset

JSON Formatter:
  Enabled:                      unset
  Indent style:                 unset
  Indent width:                 unset
  Line ending:                  unset
  Line width:                   unset
  Trailing Commas:              unset
  Expand lists:                 unset

CSS Formatter:
  Enabled:                      true
  Indent style:                 unset
  Indent width:                 unset
  Line ending:                  unset
  Line width:                   unset
  Quote style:                  unset

GraphQL Formatter:
  Enabled:                      unset
  Indent style:                 unset
  Indent width:                 unset
  Line ending:                  unset
  Line width:                   unset
  Bracket spacing:              unset
  Quote style:                  unset

Workspace:
  Open Documents:               0
```

### Configuration

```JSON
{
	"$schema": "node_modules/@biomejs/biome/configuration_schema.json",
	"assist": {
		"actions": {
			"source": {
				"organizeImports": "on",
				"useSortedAttributes": "on",
				"useSortedKeys": "on",
				"useSortedProperties": "on"
			}
		},
		"enabled": true
	},
	"css": {
		"formatter": {
			"enabled": true
		},
		"parser": {
			"tailwindDirectives": true
		}
	},
	"files": {
		"ignoreUnknown": false
	},
	"formatter": {
		"enabled": true,
		"indentStyle": "tab"
	},
	"html": {
		"experimentalFullSupportEnabled": true,
		"formatter": {
			"enabled": true,
			"indentScriptAndStyle": true
		}
	},
	"javascript": {
		"formatter": {
			"quoteStyle": "double"
		}
	},
	"linter": {
		"enabled": true,
		"rules": {
			"nursery": {
				"useSortedClasses": "on"
			},
			"recommended": true
		}
	},
	"overrides": [
		{
			"includes": ["**/*.astro", "**/*.svelte"],
			"linter": {
				"rules": {
					"correctness": {
						"noUnusedImports": "off",
						"noUnusedVariables": "off"
					},
					"style": {
						"useConst": "off",
						"useImportType": "off"
					}
				}
			}
		}
	],
	"vcs": {
		"clientKind": "git",
		"enabled": true,
		"useIgnoreFile": true
	}
}
```

### Playground link

https://biomejs.dev/playground/?indentScriptAndStyle=true&code=PABzAGMAcgBpAHAAdAA%2BAAoAIAAgAGMAbwBuAHMAdAAgAGYAcgBhAG0AZQB3AG8AcgBrACAAPQAgACIAQQBzAHQAcgBvACIAOwAKADwALwBzAGMAcgBpAHAAdAA%2BAAoACgA8AGgAMQA%2BAEgAZQBsAGwAbwAsACAAewBmAHIAYQBtAGUAdwBvAHIAawB9ACAAYQBuAGQAIABTAHYAZQBsAHQAZQAhADwALwBoADEAPgA%3D&language=svelte

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

### PR DESCRIPTION
**Pull request by dyc3**
## Summary

I threw opus at the problem and somehow, it actually did it. Usually AI can't reason about the formatter too well.

fixes #8584

## Test Plan

snapshots

### CODE REVIEW

File: crates/biome_html_formatter/src/utils/children.rs [hunk 1]
```diff
@@ -396,6 +396,27 @@ where
...
+                && let Some(last_token) = child.syntax().last_token()
+                && last_token.has_trailing_whitespace()
+            {
+                // Check if trailing trivia contains a newline
+                let has_newline = last_token
+                    .trailing_trivia()
+                    .pieces()
+                    .any(|piece| piece.is_newline());
+                if has_newline {
+                    builder.entry(HtmlChild::Newline);
```
Comment (ematipico) [inline crates/biome_html_formatter/src/utils/children.rs:418]: The tests don't cover the newline. We should add one

**Comment by Carbophile** [comment]
@dyc3 This should also apply to regular HTML elements. Apart from parity with Prettier, this quite literally isn't the same code, it renders differently in the browser.

image

File: crates/biome_html_formatter/src/utils/children.rs [hunk 2]
```diff
@@ -396,6 +396,27 @@ where
             } else {
                 builder.entry(HtmlChild::NonText(child.clone()));
             }
+
+            // Check for trailing whitespace on embedded content elements (e.g., after `}` in `{framework} `)
+            // This preserves spaces between expressions and following text content.
+            // Only apply to AnyHtmlContent (expressions), not to regular HTML elements.
+            // See: https://github.com/biomejs/biome/issues/8584
```
Comment (ematipico) [inline crates/biome_html_formatter/src/utils/children.rs]:
```suggestion
```

Probably not needed

File: crates/biome_html_formatter/src/utils/children.rs (hunk 1 above)
Comment (dyc3) [reply crates/biome_html_formatter/src/utils/children.rs:418]: I did add one that covers newlines: `crates/biome_html_formatter/tests/specs/html/svelte/whitespace/issue-8584-w-newline.svelte`

It doesn't work in the playground currently, but with https://github.com/biomejs/website/pull/3697:
image

Unless you meant newlines somewhere else?

**Comment by dyc3** [comment]
@Carbophile Good catch, thanks.

File: crates/biome_html_formatter/src/utils/children.rs (hunk 1 above)
Comment (ematipico) [reply crates/biome_html_formatter/src/utils/children.rs:418]: I missed that, thank you

File: crates/biome_html_formatter/tests/specs/prettier/html/whitespace/inline-nodes.html.snap [hunk 3]
```diff
@@ -55,7 +40,8 @@ conubia nostra, per inceptos himenaeos. Donec in ornare velit.</p>
 +  neque. Donec
 +  <a href="#"
 +    ><b>maximus</b></a
-+  >rhoncus pellentesque. Aenean purus turpis, vehicula euismod ante vel,
++  >
++  rhoncus pellentesque. Aenean purus turpis, vehicula euismod ante vel,
```
Comment (ematipico) [inline crates/biome_html_formatter/tests/specs/prettier/html/whitespace/inline-nodes.html.snap]: This seems to be a regression. It seems there's a newline after `>`

File: .changeset/evil-fans-live.md [hunk 4]
```diff
@@ -0,0 +1,10 @@
+---
```
Comment (ematipico) [inline .changeset/evil-fans-live.md:1]: The changeset should be more generic, since the fix affects other cases

File: crates/biome_html_formatter/src/utils/children.rs [hunk 5]
```diff
@@ -396,6 +396,29 @@ where
             } else {
                 builder.entry(HtmlChild::NonText(child.clone()));
             }
+
+            // Check for trailing whitespace on embedded content elements (e.g., after `}` in `{framework} `)
```
Comment (ematipico) [inline crates/biome_html_formatter/src/utils/children.rs]: The comment is wrong, because this affects other cases, as per code. In fact, it checks for elements too, not just single text expressions

File: crates/biome_html_formatter/tests/specs/prettier/html/whitespace/inline-nodes.html.snap (hunk 3 above)
Comment (dyc3) [reply crates/biome_html_formatter/tests/specs/prettier/html/whitespace/inline-nodes.html.snap]: Fixed, but it introduced another (smaller?) regression. I think long term we are going to need an entirely different approach to how html element list is formatted.
